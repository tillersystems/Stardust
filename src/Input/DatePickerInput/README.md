# DatePickerInput

### Usage

The DatePickerInput component is a text input that toggles a popover with a date picker.
The date can be updated either on the input or on the date picker displayed through a Popover
component. Date picker will close on selected date or on outside click if the date picker
was open.

```jsx
import { DatePickerInput } from '@tillersystems/stardust';
```

<!-- STORY -->

<!-- PROPS -->

### Examples

#### Uncontrolled example

```jsx
import { DatePickerInput } from '@tillersystems/stardust';

render() {
  return <DatePickerInput />
}
```

#### Controlled example

The following snippet shows how to create a controlled date picker input.

```jsx
import { DatePickerInput } from '@tillersystems/stardust';
import { DateTime } from 'luxon';

let value = DateTime.locale();

handleChange = date => value = date;

render() {
  return <DatePickerInput value={value} onChange={handleChange} />
}
```
