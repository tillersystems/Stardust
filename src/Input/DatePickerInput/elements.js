import styled from 'styled-components';

export const Container = styled.div`
  position: relative;
`;

export const DatePickerWrapper = styled.div`
  background: white;
`;

export const PopoverContainer = styled.div`
  position: absolute;
`;
