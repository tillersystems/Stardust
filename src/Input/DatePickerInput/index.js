import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { DateTime } from 'luxon';

import { DatePicker, Popover, TextInput } from '../..';
import { Container, DatePickerWrapper, PopoverContainer } from './elements';

/**
 * The DatePickerInput component is an input that toggles a popover with a date picker
 *
 * @param {luxon.DateTime} value - The currently selected date.
 * @param {Function} onChange - The callback to notify that the date changed.
 * @param {luxon.DateTime} minDate - Minimum allowed date.
 * @param {luxon.DateTime} maxDate - Maximum allowed date.
 *
 * @returns {jsx}
 */
class DatePickerInput extends PureComponent {
  // Ref to the component node for user click tracking
  ref = null;

  /** Internal state. */
  state = {
    isDatePickerOpen: false, // whether the date picker is displayed or not
    dateValue: null, // value passed down to the date picker
    textValue: null, // value displayed in the input text
    error: false, // whether input text should be displayed with an error state
  };

  /**
   * Component lifecycle method
   * Adds the listener allowing DatePicker display
   *
   */
  componentDidMount() {
    document.addEventListener('mousedown', this.handleClick, false);
    const { value, error } = this.props;
    this.setState({
      dateValue: value || DateTime.local(),
      textValue: value || DateTime.local(),
      error,
    });
  }

  /**
   * Component lifecycle method
   * Handles error value according to what the parent component passes down
   * only if the update comes from prop change and not state change
   * since state change comes from inner state update in handleInputChange and handleDateChange
   *
   * @param {Object} prevProps - The component's previous props.
   * @param {Object} prevState - The component's previous state.
   */
  componentDidUpdate(prevProps, prevState) {
    const { error: currentErrorProp } = this.props;
    const { error } = this.state;

    if (prevState.error === error && prevProps.error !== currentErrorProp) {
      this.setState({ error: currentErrorProp });
    }
  }

  /**
   * Component lifecycle method
   * Removes the listener allowing DatePicker display
   *
   */
  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleClick, false);
  }

  /**
   * Toggles visibility of the date picker.
   * Closes the date picker if the user has clicked outside this component
   *
   * @param {Object} event - event on click
   */
  handleClick = event => {
    if (!event || (event && !this.ref.contains(event.target))) {
      this.setState({
        isDatePickerOpen: false,
      });
    }
  };

  /**
   * Handles change of date from DatePicker.
   *
   * @param {Object} dateValue - clicked date on date picker component
   */
  handleDateChange = dateValue => {
    const { onChange } = this.props;
    onChange(dateValue);

    this.setState({ dateValue, textValue: dateValue, isDatePickerOpen: false, error: false });
  };

  /**
   * Handles change of value from text input.
   * If the value is valid from a Luxon point of view, it will be passed to the calendar
   * Otherwise, we set the date value for the DatePicker to the current time.
   * Thus we keep the text and date values out of sync to allow the user to update the date
   * while keeping a valid date for the DatePicker
   * Validation can be handled higher with the callback
   *
   * @param {Object} textValue - text updated on TextInput
   */
  handleInputChange = textValue => {
    const { minDate, maxDate, onChange } = this.props;
    onChange(newValue);

    this.setState({ textValue });

    const [day, month, year] = textValue.split('/').map(v => parseInt(v, 10));
    const newValue = DateTime.fromObject({ year, month, day });
    if (
      day &&
      month &&
      year &&
      newValue.isValid &&
      (!maxDate || newValue <= maxDate) &&
      (!minDate || newValue >= minDate)
    ) {
      this.setState({ dateValue: newValue, error: false });
    } else {
      this.setState({ dateValue: DateTime.local(), error: true });
    }
  };

  /**
   * Renders the component.
   *
   * @returns {jsx}
   */
  render() {
    const { fluid, minDate, maxDate } = this.props;
    const { isDatePickerOpen, error, textValue, dateValue } = this.state;

    return (
      <Container innerRef={node => (this.ref = node)}>
        <TextInput
          fluid={fluid}
          value={textValue ? textValue.toLocaleString(DateTime.DATE_SHORT) : ''}
          type="text"
          onFocus={() => this.setState({ isDatePickerOpen: true })}
          onChange={this.handleInputChange}
          error={error}
        />
        <PopoverContainer>
          <Popover active={isDatePickerOpen} arrowPositionX="6rem">
            <DatePickerWrapper>
              <DatePicker
                minDate={minDate}
                maxDate={maxDate}
                value={dateValue}
                onChange={this.handleDateChange}
              />
            </DatePickerWrapper>
          </Popover>
        </PopoverContainer>
      </Container>
    );
  }
}

const { bool, func, object } = PropTypes;
/** Prop types validation. */
DatePickerInput.propTypes = {
  fluid: bool,
  value: object,
  onChange: func,
  error: bool,
  maxDate: object,
  minDate: object,
};

/** Default props. */
DatePickerInput.defaultProps = {
  fluid: false,
  value: null,
  onChange: () => {},
  error: false,
  maxDate: null,
  minDate: null,
};

export default DatePickerInput;
