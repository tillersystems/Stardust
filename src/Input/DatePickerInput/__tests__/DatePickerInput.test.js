import React from 'react';
import * as Luxon from 'luxon';
import { fireEvent, wait, cleanup } from 'react-testing-library';

import DatePickerInput from '..';

describe('<DatePickerInput />', () => {
  const mockLocalDateTime = Luxon.DateTime.fromObject({
    year: 2018,
    month: 7,
    day: 15,
    hour: 0,
    minute: 0,
    second: 0,
    zone: 'Europe/Paris',
  });

  beforeEach(() => {
    Luxon.DateTime.local = jest.fn().mockImplementation(() => mockLocalDateTime);
  });

  afterEach(cleanup);

  test('should render without a problem', () => {
    const { container } = render(<DatePickerInput />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should select the provided date value', async () => {
    const { getByTestId, getByText } = render(<DatePickerInput value={mockLocalDateTime} />);

    const inputNode = getByTestId('input');
    fireEvent.focus(inputNode);

    await wait();
    const selectedDay = getByText('15');

    expect(selectedDay).toHaveStyleRule('border-top-left-radius', '0.4rem');
    expect(selectedDay).toHaveStyleRule('border-top-right-radius', '0.4rem');
  });

  test('should update the selected day in date picker according to input value', async () => {
    const { getByTestId, getByText } = render(<DatePickerInput value={mockLocalDateTime} />);
    const inputNode = getByTestId('input');
    fireEvent.focus(inputNode);
    await wait();

    const selectedDay = getByText('15');
    expect(selectedDay).toHaveStyleRule('border-top-left-radius', '0.4rem');

    const inputValue = '10/01/2018';
    fireEvent.change(inputNode, { target: { value: inputValue } });

    const newSelectedDay = getByText('10');
    expect(selectedDay).not.toHaveStyleRule('border-top-left-radius', '0.4rem');
    expect(newSelectedDay).toHaveStyleRule('border-top-left-radius', '0.4rem');
  });

  test('should update the input value according to selected date in date picker', async () => {
    const { getByTestId, getByText } = render(<DatePickerInput value={mockLocalDateTime} />);
    const inputNode = getByTestId('input');
    fireEvent.focus(inputNode);
    await wait();

    const selectedDay = getByText('15');
    expect(selectedDay).toHaveStyleRule('border-top-left-radius', '0.4rem');

    const newSelectedDay = getByText('21');
    fireEvent.click(newSelectedDay);
    await wait(() => expect(newSelectedDay).not.toBeInTheDocument());

    expect(inputNode.value).toEqual('7/21/2018');
  });

  test('should display input with error when input value is invalid', async () => {
    const { getByTestId, queryByTestId } = render(<DatePickerInput value={mockLocalDateTime} />);
    const inputNode = getByTestId('input');

    const statusNode = queryByTestId('status');
    expect(statusNode).not.toBeInTheDocument();

    const inputValue = '10/01/';
    fireEvent.change(inputNode, { target: { value: inputValue } });

    const errorStatusNode = getByTestId('status');
    expect(errorStatusNode).toHaveStyleRule('background', 'hsl(6,79%,65%)');
  });

  test('should display input with error when input value is out of range', async () => {
    const props = {
      value: mockLocalDateTime,
      minDate: Luxon.DateTime.fromObject({
        year: 2018,
        month: 8,
        day: 20,
        zone: 'Europe/Paris',
      }),
      maxDate: Luxon.DateTime.fromObject({
        year: 2018,
        month: 9,
        day: 20,
        zone: 'Europe/Paris',
      }),
    };
    const { getByTestId, queryByTestId } = render(<DatePickerInput {...props} />);
    const inputNode = getByTestId('input');

    const statusNode = queryByTestId('status');
    expect(statusNode).not.toBeInTheDocument();

    const inputValue = '10/07/2018';
    fireEvent.change(inputNode, { target: { value: inputValue } });

    const errorStatusNode = getByTestId('status');
    expect(errorStatusNode).toHaveStyleRule('background', 'hsl(6,79%,65%)');
  });
});
