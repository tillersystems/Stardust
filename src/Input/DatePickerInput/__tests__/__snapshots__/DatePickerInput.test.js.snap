// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DatePickerInput /> should render without a problem 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 25rem;
  height: 4rem;
  overflow: hidden;
  border-radius: 0.4rem;
  border: 1px solid hsl(207,22%,90%);
  background-color: hsl(0,100%,100%);
}

.c2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 100%;
  width: 100%;
  background-color: hsl(0,100%,100%);
  padding: 0 0.7rem;
  border-radius: 0.4rem;
  font-size: 1.6rem;
  border: none;
  text-align: left;
}

.c2::-webkit-input-placeholder {
  font-size: 1.4rem;
}

.c2::-moz-placeholder {
  font-size: 1.4rem;
}

.c2:-ms-input-placeholder {
  font-size: 1.4rem;
}

.c2::placeholder {
  font-size: 1.4rem;
}

.c2:focus,
.c2:active {
  outline: none;
}

.c0 {
  position: relative;
}

.c3 {
  position: absolute;
}

<div
  class="c0"
>
  <div
    class="c1"
    data-testid="input-container"
    width="25rem"
  >
    <input
      class="c2"
      data-testid="input"
      id=""
      placeholder=""
      tabindex="0"
      type="text"
      value="7/15/2018"
    />
  </div>
  <div
    class="c3"
  />
</div>
`;
